Move stop/start functions from utils.py into BitcoinTestFramework

jnewbery opened the pull request:
Final preparation step in #10082 before introducing the TestNode class.

This builds on top of #10555. The first 5 commits are from that PR. Without those commits, the travis build times out.

This PR moves the stop/start functions from utils.py into BitcoinTestFramework. It also moves stateful functions/variables from utils.py into BitcoinTestFramework (coverage variables, bitcoind_processes dict, mocktime functions). It also does some general tidyup of test_framework.py and utils.py.

This touches a few individual test cases, but those changes are very minor. The important changes are in util.py and test_framework.py. The final commit is a code move only.

jnewbery committed:
[tests] Move stop_node and start_node methods to BitcoinTestFramework

This commit moves functions start_node, start_nodes, stop_node and
stop_nodes functions into the BitcoinTestFramework class. It also moves
the bitcoind_processes dict and coverage variables into BitcoinTestFramework.

jnewbery committed:
[tests] fix flake8 warnings in test_framework.py and util.py

jnewbery committed:
[tests] remove unused imports from utils.py

jnewbery committed:
[tests] move mocktime property and functions to BitcoinTestFramework

jnewbery committed:
[tests] reorganize utils.py module (code move only)

This commit re-organizes the utils.py module into logical sections.

jnewbery commented:
@jimmysong @kallewoof @sdaftuar - you've all mentioned that you're happy to help move #10082 forward. Any review of this and/or #10555 would be appreciated.

jimmysong commented on test/functional/zmq_test.py:
```diff
@@ -38,57 +38,55 @@ def setup_nodes(self):
 
-        for x in range(0,n):
-            assert_equal(genhashes[x], zmqHashes[x]) #blockhash from generate must be equal to the hash received over zmq
+        for x in range(0, n):
```
nit: `range(n)` is better than `range(0, n)`

jnewbery replied:
Agree. Changed in #10555

jimmysong commented on test/functional/zmq_test.py:
```diff
@@ -6,12 +6,23 @@
                                  )
 
+def wait_for_multipart(socket, timeout=60):
+    while True:
```
This might be easier to read if made into a for loop:

    interval = 0.1
    for _ in range(int(timeout/interval)):
        try:
            return socket.recv_multipart(zmq.NOBLOCK)
        except zmq.error.Again:
            time.sleep(interval)
    raise AssertionError("Timed out waiting for zmq message")

jnewbery replied:
Yes, that's better. I've made the change in #10555, and will rebase this PR off that once you've finished your review.

jimmysong commented on test/functional/fundrawtransaction.py:
```diff
@@ -452,7 +452,7 @@ def run_test(self):
         self.stop_node(3)
         self.nodes[1].encryptwallet("test")
-        bitcoind_processes[1].wait(timeout=BITCOIND_PROC_WAIT_TIMEOUT)
+        self.bitcoind_processes[1].wait(60)
```
Why take out the constant here?

jnewbery replied:
Because I've moved the `BITCOIND_PROC_WAIT_TIMEOUT` constant into test_framework.py. The next PR (introduce TestNode) will remove `bitcoind_processes` dict entirely, so for this intermediate PR rather than add an import which I'll revert in the next PR, I've just changed this to a hard-coded value.

MarcoFalke replied:
Please keep the arg named. I'd really prefer if all primitive types are named in our python code.

jnewbery replied:
Done. I've also restored the use of a constant here (at least until the next PR in this series is merged).

MarcoFalke replied:
Are you sure this was fixed? GitHub didn't collapse the comments.

jnewbery replied:
Oops. Fixed but not pushed.

Should be good now.

jimmysong commented on test/functional/test_framework/test_framework.py:
```diff
@@ -14,6 +14,7 @@
 import sys
 import tempfile
 import time
+import traceback
```
Is this being used?

jnewbery replied:
Here: https://github.com/bitcoin/bitcoin/blob/9c248e39f2807a7f89e555e99cc55cb2de1ad335/test/functional/test_framework/test_framework.py#L194

I must have missed out this import in a previous commit. Makes sense to fix it in this cleanup commit.

jimmysong commented on test/functional/bip9-softforks.py:
```diff
@@ -15,16 +15,17 @@
 test that enforcement has triggered
 """
+from io import BytesIO
+import shutil
```
How did this work before? Was it one of the import *'s?

jnewbery replied:
Yes, this was being indirectly imported from utils.py :( there are still lots of indirect imports in functional tests, which makes refactors like this seem more disruptive than they actually are.

jimmysong reviewed (commented):
utAck f7acc7adea731589aec8a7d9e16cc9069e24b5a9

Added a few nits and questions. Testing next.

jnewbery commented:
@jimmysong - Thanks for the quick review! Responses inline.

jimmysong commented:
Great, please let me know when you rebase to #10555 and I'll test it.

jnewbery commented:
rebased on #10555

jimmysong commented:
Tested ACK e1e3dd074b71503d8f74651bf2a3cfd9e55a4d2b

achow101 commented:
utACK e1e3dd074b71503d8f74651bf2a3cfd9e55a4d2b

kallewoof commented on test/functional/test_framework/test_framework.py:
```diff
@@ -207,16 +203,82 @@ def main(self):
+                rpcs.append(self.start_node(i, dirname, extra_args[i], rpchost, timewait=timewait, binary=binary[i]))
+        except:
+            # If one node failed to start, stop the others
+            self.stop_nodes()
```
I have a hard time understanding how this will work. `self.nodes` seems to be set to `self.start_nodes(...)` which means at this point, `self.nodes` is not set at all, and `self.stop_nodes()` will do nothing. Am I missing something?

achow101 replied:
I don't think you are. Instead of returning `rpcs`, I think this function should appent to `self.nodes` directly.

jnewbery replied:
Good spot @kallewoof - you're not missing anything at all. This is obviously wrong :(

Once #10082 is fully merged, I'd like to improve the way the TestNode objects are handled by the BitcoinTestFramework class. There should be separate methods `add_nodes()`, which would instantiate new TestNode objects, and `start_node()`, which would start the bitcoind process. I don't think `start_node()` or `start_nodes()` should be responsible for updating the `self.nodes` list, and individual tests shouldn't normally have to concern themselves with updating the list.

I also think that eventually `self.nodes` should be a dictionary.

For now, I've added a fixup commit that appends the newly created nodes to self.nodes, then calls self.stop_nodes(), and then clears the list. It's a little hacky, but it should restore the functionality here until #10082 is fully merged. Can you take a look and let me know what you think?

kallewoof replied:
@jnewbery Fix looks good to me!

kallewoof reviewed (approved):
Tested ACK e1e3dd0

jnewbery commented:
Needed rebase because the new wallet-encryption test used functions that are moved into the BitcoinTestFramework class by the PR.

jnewbery commented:
Rebased

laanwj commented:
Needs rebase after #10533

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -187,7 +183,7 @@ def main(self):
                     try:
                         with open(fn, 'r') as f:
-                            print("From" , fn, ":")
+                            print("From", fn, ":")
```
Not sure how to feel about the whitespace changes. This might discourage or distract reviewers. Maybe we could make those in a separate somewhat larger pull?

jnewbery commented:
rebased (twice - since master changed while I was rebasing 🙁 )

> Not sure how to feel about the whitespace changes. This might discourage or distract reviewers.

I include style fixes for the modules I'm touching in PRs for a few reasons:
- I like my commits to run clean through a linter. It's the best way I know to make sure I'm not introducing new style nits.
- no-one likes reviewing style-change-only PRs, so unless they're included in larger PRs, style tends to degrade over time
- it's less disruptive than project-wide style PRs in terms of rebasing, since the module is being updated anyway.

It doesn't appear that this has discouraged reviewers - this PR has three ACKs and a comment from another reviewer (you).

The style changes are contained within separate commits so if you feel really strongly about this I can remove them. However, I'd prefer not to since I've already rebased this PR five times.

MarcoFalke commented:
Sorry, this needs rebase again. :(

jnewbery commented:
rebased

MarcoFalke commented:
Well, the issue with including whitespace fixes is that it makes it harder
to backport and also harder to keep mergeable. So the rebases will
introduce errors (see below). Actually I would be fine with a cleanup in
test that fixes whitespace and is a `git diff --ignore-all-space` empty
diff.

utACK with 05b8c081b435e87b08335e9f9b62a55fa1d48ecc, lets leave this as is
and fix two nits:

* rpc_auth_pair does not exist previously and is never used, please remove
* You want to move the `assert_equal(return_code, 0)` down by one line,
such that the process is deleted even if the return code does not match
* wait_for_node_exit should probably come with a default value for timeout,
but this should be done i a follow a later pull.

On Thu, Jun 29, 2017 at 12:59 PM, John Newbery <notifications@github.com>
wrote:

> rebased
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/10556#issuecomment-311933188>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGGmv1fsdTJNMgo7Qfy6NzuthY9Ulju5ks5sI4OMgaJpZM4N0NGy>
> .
>

jnewbery committed:
[tests] fix nits.

Thanks to Marco Falke.

jnewbery commented:
Fixed the first two nits. Thanks for catching

> wait_for_node_exit should probably come with a default value for timeout,
but this should be done i a follow a later pull.

Right - I think that's completely orthogonal to this pull. It was introduced in 176c021d085f5a45bc9e038e760942aa648dd797
